contracts/random/PRNG.sol

```diff
@@ -38,31 +41,32 @@ library PRNG {
     }
 
     /**
-    @dev Hashes seed||counter, placing it in the entropy word, and resets the
-    remaining bits to 256. Increments the counter BEFORE the refill (ie 0 is
-    never used) as this simplifies round-tripping with store() and loadSource()
-    because the stored counter state is the same as the one used for deriving
-    the entropy pool.
+    @dev Computes the next PRN in entropy using a lag-1 multiply-with-carry
+    algorithm and resets the remaining bits to 128.
+    `nextNumber = (factor * number + carry) mod 2**128`
+    `nextCarry  = (factor * number + carry) //  2**128`
      */
     function _refill(Source src) private pure {
         assembly {
-            let ctr := add(src, COUNTER)
-            mstore(ctr, add(1, mload(ctr)))
-            mstore(add(src, ENTROPY), keccak256(src, 0x40))
-            mstore(add(src, REMAIN), 256)
+            let carryAndNumber := mload(src)
+            let rand := and(carryAndNumber, 0xffffffffffffffffffffffffffffffff)
+            let carry := shr(128, carryAndNumber)
+            let tmp := add(mul(MWC_FACTOR, rand), carry)
+            mstore(add(src, REMAIN), 128)
+            mstore(src, tmp)
         }
     }
 
     /**
-    @notice Returns the specified number of bits <= 256 from the Source.
+    @notice Returns the specified number of bits <= 128 from the Source.
```

Comment:
What do you think about changing it to a `uint8` in this case? I'm sure there's a downside, but not sure if it outweighs the stronger type safety. I wanted to for the original, but needed to accommodate the 256-bit read.

Reply:
Definitely. I also intended to change all `uint256 bits` in both files but opted to keep the PR as slim as possbile for now.

Reply:
The only potential problem that I see with this is that both PRNG libraries don't share a common interface any longer. Also I think the gain in safety is negligible since `uint8` still includes a wide range of bit widths (129-255) for which `PRNG.read` calls will revert.

I spotted some `uint256 bits` in both library that I wanted to change to `uint16` to be consistent . However, I opted not to, to keep the PR slim. I'd open a separate one, if you agree on the change.

Reply:
Yeah I agree with both ideas, keeping the signature the same, and changing to `uint16` in a later PR.